File: services/api/app/domain/friendship.py
"""The friend graph, as a state machine. F03 and F04.

One edge between two people, and the rule that makes it an edge rather than a
claim: **the addressee decides**. A friendship exists when the person who was
asked said yes. Nothing else creates one.

That sentence is the whole reason this module is pure and separate. Written the
ordinary way -- a `friends` table an endpoint inserts into -- "add friend" is a
write the requester performs on somebody else's social graph, and every later
reader (who may see the trip, who may be invited, whose name autocompletes)
inherits a relationship the other person never agreed to. Section 9's confused
deputy, wearing a friendly word.

So there is no `add_friend`. There is `open_request`, which creates a PENDING
edge and grants nothing, and `decide`, which only the addressee may call.
`are_friends` is derived from the resulting state, never stored as its own
truth -- the same shape as invariant 3 for money: the relationship is
recomputable from the events that produced it.

## The four states are the spec's, exactly

`feature_list.md` F04 lists pending, accepted, declined, blocked. They are not
four flags on a row; they are the four places an edge can rest, and the legal
moves between them are the table below.

    PENDING --accept--> ACCEPTED      addressee only
    PENDING --decline-> DECLINED      addressee only
    PENDING --block---> BLOCKED       addressee only
    ACCEPTED -block---> BLOCKED       either party -- friendship can end badly
    DECLINED -----> (reopen)          a new request, not a transition

    BLOCKED --unblock-> DECLINED      the blocker only

DECLINED is deliberately not terminal: being turned down once is not a life
sentence, and the alternative is a product where a mistyped tap permanently
removes somebody from your reachable set. BLOCKED is terminal *for the person
who was blocked*: `open_request` refuses to reopen a blocked edge and refuses
it without saying who blocked whom (see `BLOCKED_IS_SILENT`), so from that
side the wall has no door in it. The person who put the wall up may take it
down -- `unblock`, ADR-0023 §2.3.3 -- and what they get back is DECLINED, not
a friendship: undoing a block must not silently re-create a relationship the
other person agreed to under different circumstances. Blocking twice is the
same wall, not an error.

## Pair keys

`(A asks B)` and `(B asks A)` are the same edge, so state has to be keyed by an
unordered pair. `pair_key` sorts the two ids, which is what lets the database
enforce "at most one live edge per pair" with one unique index instead of two
rows racing each other into a friendship nobody can undo. The SQL index says
the same thing in `least()/greatest()`; keep the two spellings agreeing.

Pure: `dict` in, `dict` out. Raises `FriendshipError`. Imports nothing from
`app.db`, `app.api`, sqlalchemy or fastapi -- `tests/test_import_boundary.py`
parses this file to check that, so the check is not a promise.
"""

from __future__ import annotations

from enum import StrEnum

__all__ = [
    "BLOCKED_IS_SILENT",
    "Decision",
    "FriendState",
    "FriendshipError",
    "are_friends",
    "decide",
    "is_live_edge",
    "open_block",
    "open_request",
    "pair_key",
    "unblock",
]


class FriendState(StrEnum):
    """The four rest states of one edge. F04's list, unchanged."""

    PENDING = "pending"
    ACCEPTED = "accepted"
    DECLINED = "declined"
    BLOCKED = "blocked"


class Decision(StrEnum):
    """What an addressee may answer. Accepting is one of three, not the default."""

    ACCEPT = "accept"
    DECLINE = "decline"
    BLOCK = "block"


#: A blocked person must not be able to tell blocking apart from "no account",
#: "never answered", or "already asked". If the refusal named the block, the
#: block would announce itself to exactly the person it protects somebody from,
#: and the product would be teaching users that blocking is unsafe. So
#: `open_request` raises the same code for a blocked edge as for a duplicate
#: one, and the caller must map both to the same HTTP answer.
BLOCKED_IS_SILENT = "REQUEST_NOT_OPEN"


class FriendshipError(Exception):
    """A refusal with a stable code. The code is the contract; the text is not."""

    def __init__(self, code: str):
        super().__init__(code)
        self.code = code


#: States in which an edge still occupies the pair -- a new request may not be
#: opened alongside one of these. DECLINED is absent on purpose: that is the
#: state a person may ask again from.
_LIVE = (FriendState.PENDING, FriendState.ACCEPTED, FriendState.BLOCKED)


def pair_key(a: str, b: str) -> tuple[str, str]:
    """The unordered pair, canonically ordered.

    Sorted as strings, matching what `least()/greatest()` do to the same two
    values in the partial unique index. Two spellings of one rule is a risk;
    they are kept adjacent in the migration's comment for that reason.
    """
    if not a or not b:
        raise FriendshipError("PERSON_REQUIRED")
    if a == b:
        # Nobody friends themselves. Allowed, this becomes a self-edge that
        # every "who are my friends" query has to special-case forever.
        raise FriendshipError("SELF_EDGE")
    return (a, b) if a < b else (b, a)


def is_live_edge(state: str) -> bool:
    """True when this state still occupies the pair against a new request."""
    return FriendState(state) in _LIVE


def open_request(
    *, requester_id: str, addressee_id: str, existing: dict | None = None
) -> dict:
    """Open a PENDING edge from requester to addressee.

    `existing` is the current edge for this pair, if any, as previously
    returned by this module. Passing it is how the caller proves it looked --
    the parameter is not optional in spirit, only in signature, and the
    repository always reads before writing.

    Grants nothing. A PENDING edge is a question, and `are_friends` is False
    for it.
    """
    pair_key(requester_id, addressee_id)  # raises on self-edge / empty

    if existing is not None and is_live_edge(existing["state"]):
        # One code for pending, accepted and blocked alike. See
        # BLOCKED_IS_SILENT: distinguishing them here is what would leak the
        # block. "You already have a live edge with this person" is true in
        # all three cases and reveals nothing new to a legitimate caller,
        # because a legitimate caller can already see their own pending and
        # accepted edges by listing them.
        raise FriendshipError(BLOCKED_IS_SILENT)

    return {
        "requester_id": requester_id,
        "addressee_id": addressee_id,
        "state": str(FriendState.PENDING),
        "pair": list(pair_key(requester_id, addressee_id)),
    }


def decide(*, edge: dict, actor_id: str, decision: str) -> dict:
    """Answer a request. **Only the addressee may accept.**

    This function is the consent gate for the whole feature. The permission
    table refuses an actor who is not the addressee before the request reaches
    here, and this refuses it again on facts read from the row itself. Two
    layers because they fail differently: the table is data somebody could edit
    without reading this file, and this is logic somebody could call without
    going through a route.

    Blocking is the one answer either party may give, and the only one legal
    from ACCEPTED -- a friendship that has to end does not need the other
    side's permission to end.
    """
    answer = Decision(decision)
    state = FriendState(edge["state"])
    requester_id = edge["requester_id"]
    addressee_id = edge["addressee_id"]

    if actor_id not in (requester_id, addressee_id):
        raise FriendshipError("NOT_A_PARTY")

    if answer is Decision.BLOCK:
        if state is FriendState.BLOCKED:
            raise FriendshipError("ALREADY_BLOCKED")
        # Either party, from PENDING or ACCEPTED. A requester who realises they
        # asked the wrong person is also using this door.
        return {**edge, "state": str(FriendState.BLOCKED)}

    # Accept and decline are answers to a question, so they need a question.
    if state is not FriendState.PENDING:
        raise FriendshipError("NOT_PENDING")

    if actor_id != addressee_id:
        # The line the acceptance criteria mutate. A requester accepting their
        # own request is the entire failure mode this feature exists to
        # prevent: it is "add friend" with extra steps.
        raise FriendshipError("ONLY_ADDRESSEE_MAY_ANSWER")

    return {
        **edge,
        "state": str(
            FriendState.ACCEPTED if answer is Decision.ACCEPT else FriendState.DECLINED
        ),
    }


def open_block(*, blocker_id: str, addressee_id: str, existing: dict | None) -> dict:
    """Block somebody, whether or not an edge exists yet (ADR-0023 §2.3).

    Three starting points, one ending: no edge at all (a stranger, or somebody
    whose earlier request was declined), a live edge, and an edge that is
    already a block. The first writes a fresh `blocked` row whose requester is
    the blocker -- there was no question, so there is nobody to have asked;
    the second is `decide(BLOCK)`, which either party may call; the third is
    the same wall and answers `ALREADY_BLOCKED` for the caller to read as
    «done», not as a failure.

    `decided_by_id` is the blocker in every case. `blocker_of` reads that
    field, and `unblock` is the only door it opens.
    """
    pair_key(blocker_id, addressee_id)  # raises on self-edge / empty
    if existing is None or FriendState(existing["state"]) is FriendState.DECLINED:
        return {
            "requester_id": blocker_id,
            "addressee_id": addressee_id,
            "state": str(FriendState.BLOCKED),
            "decided_by_id": blocker_id,
            "pair": list(pair_key(blocker_id, addressee_id)),
        }
    return decide(edge=existing, actor_id=blocker_id, decision=str(Decision.BLOCK))


def unblock(*, edge: dict | None, actor_id: str) -> dict:
    """Lift a block. Only whoever put it up, and never back into friendship.

    `NOT_BLOCKED` for an edge in any other state, so a client that lost track
    cannot turn an accepted friendship into a declined one by pressing the
    wrong button. `ONLY_BLOCKER_MAY_UNBLOCK` for the person on the wrong side
    of the wall -- that predicate is the whole point of the feature, and it is
    proved here on the row as well as in the permission table.
    """
    if edge is None or FriendState(edge["state"]) is not FriendState.BLOCKED:
        raise FriendshipError("NOT_BLOCKED")
    if edge.get("decided_by_id") != actor_id:
        raise FriendshipError("ONLY_BLOCKER_MAY_UNBLOCK")
    return {**edge, "state": str(FriendState.DECLINED), "decided_by_id": actor_id}


def are_friends(edge: dict | None) -> bool:
    """Derived, never stored. Only ACCEPTED is a friendship."""
    if edge is None:
        return False
    return FriendState(edge["state"]) is FriendState.ACCEPTED

File: services/core/internal/domain/friendship/friendship.go
// Package friendship ports app.domain.friendship (F03, F04): the friend graph
// as a state machine in which only the addressee turns a request into a
// friendship.
//
// Parity, not correctness, is the contract (ADR-0029 §2.4): oracle_test.go
// replays testdata/python_*.json, rendered by
// scripts/render_domain_w2_goldens.py from the real module in the parity API
// image. Behaviour that looks surprising is Python's and is kept; the
// function comments say where.
//
// # Values
//
// An Edge is the dict the service and the repository build. DecidedByID nil is
// Python's None and also an absent key: nothing in the module tells the two
// apart. Pair nil is an absent "pair" key. Strings are Python str values, so
// they must be valid UTF-8; state and decision words are checked the way the
// StrEnum lookups check them, and an unknown word is a *ValueError whose
// message is Python's, repr and all.
package friendship

import (
	"slices"
	"strconv"
	"strings"
	"unicode"
)

// The four rest states of one edge (FriendState).
const (
	StatePending  = "pending"
	StateAccepted = "accepted"
	StateDeclined = "declined"
	StateBlocked  = "blocked"
)

// What an addressee may answer (Decision).
const (
	DecisionAccept  = "accept"
	DecisionDecline = "decline"
	DecisionBlock   = "block"
)

// BlockedIsSilent is BLOCKED_IS_SILENT: the one code open_request gives for a
// pending, accepted or blocked edge alike, so a block never names itself.
const BlockedIsSilent = "REQUEST_NOT_OPEN"

// Codes FriendshipError carries besides BlockedIsSilent.
const (
	CodePersonRequired         = "PERSON_REQUIRED"
	CodeSelfEdge               = "SELF_EDGE"
	CodeNotAParty              = "NOT_A_PARTY"
	CodeAlreadyBlocked         = "ALREADY_BLOCKED"
	CodeNotPending             = "NOT_PENDING"
	CodeOnlyAddresseeMayAnswer = "ONLY_ADDRESSEE_MAY_ANSWER"
	CodeNotBlocked             = "NOT_BLOCKED"
	CodeOnlyBlockerMayUnblock  = "ONLY_BLOCKER_MAY_UNBLOCK"
)

var (
	states    = [...]string{StatePending, StateAccepted, StateDeclined, StateBlocked}
	decisions = [...]string{DecisionAccept, DecisionDecline, DecisionBlock}
	// live is _LIVE: the states that still occupy the pair.
	live = [...]string{StatePending, StateAccepted, StateBlocked}
)

// States returns FriendState's values in declaration order.
func States() []string { return slices.Clone(states[:]) }

// Decisions returns Decision's values in declaration order.
func Decisions() []string { return slices.Clone(decisions[:]) }

// LiveStates returns _LIVE in declaration order.
func LiveStates() []string { return slices.Clone(live[:]) }

// FriendshipError is Python's FriendshipError: `str(exc)` is the code.
type FriendshipError struct {
	Code string
}

func (e *FriendshipError) Error() string { return e.Code }

func refuse(code string) error { return &FriendshipError{Code: code} }

// ValueError is the ValueError `FriendState(value)` or `Decision(value)`
// raises for a word that is not a member.
type ValueError struct {
	Message string
}

func (e *ValueError) Error() string { return e.Message }

// Edge is the edge dict. See the package comment for nil.
type Edge struct {
	RequesterID string
	AddresseeID string
	State       string
	DecidedByID *string
	Pair        []string
}

// clone copies the edge so a returned dict never aliases the caller's, the
// way `{**edge}` does not.
func (e Edge) clone() Edge {
	out := e
	out.Pair = slices.Clone(e.Pair)
	if e.DecidedByID != nil {
		decider := *e.DecidedByID
		out.DecidedByID = &decider
	}
	return out
}

func friendState(value string) (string, error) {
	if slices.Contains(states[:], value) {
		return value, nil
	}
	return "", &ValueError{Message: reprStr(value) + " is not a valid FriendState"}
}

func decision(value string) (string, error) {
	if slices.Contains(decisions[:], value) {
		return value, nil
	}
	return "", &ValueError{Message: reprStr(value) + " is not a valid Decision"}
}

// PairKey is pair_key: the unordered pair, sorted as Python sorts str. UTF-8
// byte order is code point order, so Go's < agrees.
func PairKey(a, b string) ([2]string, error) {
	if a == "" || b == "" {
		return [2]string{}, refuse(CodePersonRequired)
	}
	if a == b {
		return [2]string{}, refuse(CodeSelfEdge)
	}
	if a < b {
		return [2]string{a, b}, nil
	}
	return [2]string{b, a}, nil
}

// IsLiveEdge is is_live_edge.
func IsLiveEdge(state string) (bool, error) {
	member, err := friendState(state)
	if err != nil {
		return false, err
	}
	return slices.Contains(live[:], member), nil
}

// OpenRequest is open_request. Only existing.State is read.
func OpenRequest(requesterID, addresseeID string, existing *Edge) (Edge, error) {
	pair, err := PairKey(requesterID, addresseeID)
	if err != nil {
		return Edge{}, err
	}
	if existing != nil {
		occupied, err := IsLiveEdge(existing.State)
		if err != nil {
			return Edge{}, err
		}
		if occupied {
			return Edge{}, refuse(BlockedIsSilent)
		}
	}
	return Edge{
		RequesterID: requesterID,
		AddresseeID: addresseeID,
		State:       StatePending,
		Pair:        []string{pair[0], pair[1]},
	}, nil
}

// Decide is decide. The decision word is checked before the state word, and
// nothing checks that the edge is not a self-edge: a row naming one person
// twice lets that person accept.
func Decide(edge Edge, actorID, decisionWord string) (Edge, error) {
	answer, err := decision(decisionWord)
	if err != nil {
		return Edge{}, err
	}
	state, err := friendState(edge.State)
	if err != nil {
		return Edge{}, err
	}
	if actorID != edge.RequesterID && actorID != edge.AddresseeID {
		return Edge{}, refuse(CodeNotAParty)
	}
	out := edge.clone()
	if answer == DecisionBlock {
		if state == StateBlocked {
			return Edge{}, refuse(CodeAlreadyBlocked)
		}
		out.State = StateBlocked
		return out, nil
	}
	if state != StatePending {
		return Edge{}, refuse(CodeNotPending)
	}
	if actorID != edge.AddresseeID {
		return Edge{}, refuse(CodeOnlyAddresseeMayAnswer)
	}
	if answer == DecisionAccept {
		out.State = StateAccepted
	} else {
		out.State = StateDeclined
	}
	return out, nil
}

// OpenBlock is open_block. With no edge or a declined one it writes a fresh
// blocked edge owned by the blocker. Otherwise it is Decide(block), which
// keeps whatever decided_by_id the existing edge carried: despite the Python
// docstring, the blocker is not written there on that path.
func OpenBlock(blockerID, addresseeID string, existing *Edge) (Edge, error) {
	pair, err := PairKey(blockerID, addresseeID)
	if err != nil {
		return Edge{}, err
	}
	fresh := existing == nil
	if !fresh {
		state, err := friendState(existing.State)
		if err != nil {
			return Edge{}, err
		}
		fresh = state == StateDeclined
	}
	if fresh {
		decider := blockerID
		return Edge{
			RequesterID: blockerID,
			AddresseeID: addresseeID,
			State:       StateBlocked,
			DecidedByID: &decider,
			Pair:        []string{pair[0], pair[1]},
		}, nil
	}
	return Decide(*existing, blockerID, DecisionBlock)
}

// Unblock is unblock.
func Unblock(edge *Edge, actorID string) (Edge, error) {
	if edge == nil {
		return Edge{}, refuse(CodeNotBlocked)
	}
	state, err := friendState(edge.State)
	if err != nil {
		return Edge{}, err
	}
	if state != StateBlocked {
		return Edge{}, refuse(CodeNotBlocked)
	}
	if edge.DecidedByID == nil || *edge.DecidedByID != actorID {
		return Edge{}, refuse(CodeOnlyBlockerMayUnblock)
	}
	out := edge.clone()
	out.State = StateDeclined
	decider := actorID
	out.DecidedByID = &decider
	return out, nil
}

// AreFriends is are_friends.
func AreFriends(edge *Edge) (bool, error) {
	if edge == nil {
		return false, nil
	}
	state, err := friendState(edge.State)
	if err != nil {
		return false, err
	}
	return state == StateAccepted, nil
}

// reprStr is CPython 3.12's repr(str), as the enum's ValueError prints it.
func reprStr(s string) string {
	quote := byte('\'')
	if strings.IndexByte(s, '\'') >= 0 && strings.IndexByte(s, '"') < 0 {
		quote = '"'
	}
	var b strings.Builder
	b.WriteByte(quote)
	for _, r := range s {
		switch {
		case r == rune(quote) || r == '\\':
			b.WriteByte('\\')
			b.WriteRune(r)
		case r == '\t':
			b.WriteString(`\t`)
		case r == '\n':
			b.WriteString(`\n`)
		case r == '\r':
			b.WriteString(`\r`)
		case isPrintable(r):
			b.WriteRune(r)
		case r <= 0xFF:
			b.WriteString(`\x` + padHex(r, 2))
		case r <= 0xFFFF:
			b.WriteString(`\u` + padHex(r, 4))
		default:
			b.WriteString(`\U` + padHex(r, 8))
		}
	}
	b.WriteByte(quote)
	return b.String()
}

// isPrintable is Py_UNICODE_ISPRINTABLE: everything but categories C* and Z*,
// with the ASCII space kept. Go 1.23 and CPython 3.12 both read Unicode 15.0;
// oracle_test.go checks every code point.
func isPrintable(r rune) bool {
	if r < 0x7F {
		return r >= 0x20
	}
	return r != 0x7F && unicode.IsPrint(r)
}

func padHex(r rune, width int) string {
	digits := strconv.FormatInt(int64(r), 16)
	if len(digits) < width {
		digits = strings.Repeat("0", width-len(digits)) + digits
	}
	return digits
}
